File: cpp_backup/filters/morph_filters.cpp
#include "filters/morph_filters.h"
#include <opencv2/imgproc.hpp>

// ---------------------------------------------------------------------------
// Dilation & Erosion
// ---------------------------------------------------------------------------
QVector<PFMSetting> DilationFilter::defineSettings() const {
    return {
        { "ksize", "Kernel Size", SettingType::Integer, 3, QVariant(), 1, 31, 1, 31, 2 },
        { "iterations", "Iterations", SettingType::Integer, 1, QVariant(), 1, 10, 1, 10, 1 }
    };
}
cv::Mat DilationFilter::process(const cv::Mat& image) {
    int ksize = get("ksize").toInt();
    int iters = get("iterations").toInt();
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(ksize, ksize));
    cv::Mat out;
    cv::dilate(image, out, kernel, cv::Point(-1,-1), iters);
    return out;
}

QVector<PFMSetting> ErosionFilter::defineSettings() const {
    return {
        { "ksize", "Kernel Size", SettingType::Integer, 3, QVariant(), 1, 31, 1, 31, 2 },
        { "iterations", "Iterations", SettingType::Integer, 1, QVariant(), 1, 10, 1, 10, 1 }
    };
}
cv::Mat ErosionFilter::process(const cv::Mat& image) {
    int ksize = get("ksize").toInt();
    int iters = get("iterations").toInt();
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(ksize, ksize));
    cv::Mat out;
    cv::erode(image, out, kernel, cv::Point(-1,-1), iters);
    return out;
}

// ---------------------------------------------------------------------------
// Advanced Morph
// ---------------------------------------------------------------------------
QVector<PFMSetting> OpeningFilter::defineSettings() const {
    return { { "ksize", "Kernel Size", SettingType::Integer, 3, QVariant(), 1, 31, 1, 31, 2 } };
}
cv::Mat OpeningFilter::process(const cv::Mat& image) {
    int ksize = get("ksize").toInt();
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(ksize, ksize));
    cv::Mat out;
    cv::morphologyEx(image, out, cv::MORPH_OPEN, kernel);
    return out;
}

QVector<PFMSetting> ClosingFilter::defineSettings() const {
    return { { "ksize", "Kernel Size", SettingType::Integer, 3, QVariant(), 1, 31, 1, 31, 2 } };
}
cv::Mat ClosingFilter::process(const cv::Mat& image) {
    int ksize = get("ksize").toInt();
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(ksize, ksize));
    cv::Mat out;
    cv::morphologyEx(image, out, cv::MORPH_CLOSE, kernel);
    return out;
}

QVector<PFMSetting> MorphGradientFilter::defineSettings() const {
    return { { "ksize", "Kernel Size", SettingType::Integer, 3, QVariant(), 1, 31, 1, 31, 2 } };
}
cv::Mat MorphGradientFilter::process(const cv::Mat& image) {
    int ksize = get("ksize").toInt();
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(ksize, ksize));
    cv::Mat out;
    cv::morphologyEx(image, out, cv::MORPH_GRADIENT, kernel);
    return out;
}

QVector<PFMSetting> TopHatFilter::defineSettings() const {
    return { { "ksize", "Kernel Size", SettingType::Integer, 9, QVariant(), 1, 101, 1, 101, 2 } };
}
cv::Mat TopHatFilter::process(const cv::Mat& image) {
    int ksize = get("ksize").toInt();
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(ksize, ksize));
    cv::Mat out;
    cv::morphologyEx(image, out, cv::MORPH_TOPHAT, kernel);
    return out;
}

QVector<PFMSetting> BlackHatFilter::defineSettings() const {
    return { { "ksize", "Kernel Size", SettingType::Integer, 9, QVariant(), 1, 101, 1, 101, 2 } };
}
cv::Mat BlackHatFilter::process(const cv::Mat& image) {
    int ksize = get("ksize").toInt();
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(ksize, ksize));
    cv::Mat out;
    cv::morphologyEx(image, out, cv::MORPH_BLACKHAT, kernel);
    return out;
}

// ---------------------------------------------------------------------------
// Cross & Ellipse
// ---------------------------------------------------------------------------
cv::Mat DilateCrossFilter::process(const cv::Mat& image) {
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_CROSS, cv::Size(5, 5));
    cv::Mat out;
    cv::dilate(image, out, kernel, cv::Point(-1,-1), 1);
    return out;
}

cv::Mat ErodeCrossFilter::process(const cv::Mat& image) {
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_CROSS, cv::Size(5, 5));
    cv::Mat out;
    cv::erode(image, out, kernel, cv::Point(-1,-1), 1);
    return out;
}

cv::Mat DilateEllipseFilter::process(const cv::Mat& image) {
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(5, 5));
    cv::Mat out;
    cv::dilate(image, out, kernel, cv::Point(-1,-1), 1);
    return out;
}

cv::Mat ErodeEllipseFilter::process(const cv::Mat& image) {
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(5, 5));
    cv::Mat out;
    cv::erode(image, out, kernel, cv::Point(-1,-1), 1);
    return out;
}

File: python/app/filters/morph_filters.py
"""Morphological Filters (Erosion, Dilation, Opening, Closing)."""

import cv2
import numpy as np
from typing import List

from app.filters.base_filter import ImageFilter
from app.pfm import PFMSetting, SettingType


class DilationFilter(ImageFilter):
    @property
    def name(self) -> str: return "Dilation"
    @property
    def category(self) -> str: return "Morphological"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("ksize", "Kernel Size", SettingType.INTEGER, 3, 1, 31, 2),
            PFMSetting("iterations", "Iterations", SettingType.INTEGER, 1, 1, 10, 1)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        ksize = self.get("ksize")
        iters = self.get("iterations")
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (ksize, ksize))
        return cv2.dilate(image, kernel, iterations=iters)


class ErosionFilter(ImageFilter):
    @property
    def name(self) -> str: return "Erosion"
    @property
    def category(self) -> str: return "Morphological"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("ksize", "Kernel Size", SettingType.INTEGER, 3, 1, 31, 2),
            PFMSetting("iterations", "Iterations", SettingType.INTEGER, 1, 1, 10, 1)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        ksize = self.get("ksize")
        iters = self.get("iterations")
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (ksize, ksize))
        return cv2.erode(image, kernel, iterations=iters)


class OpeningFilter(ImageFilter):
    @property
    def name(self) -> str: return "Opening"
    @property
    def category(self) -> str: return "Morphological"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("ksize", "Kernel Size", SettingType.INTEGER, 3, 1, 31, 2)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        ksize = self.get("ksize")
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (ksize, ksize))
        return cv2.morphologyEx(image, cv2.MORPH_OPEN, kernel)


class ClosingFilter(ImageFilter):
    @property
    def name(self) -> str: return "Closing"
    @property
    def category(self) -> str: return "Morphological"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("ksize", "Kernel Size", SettingType.INTEGER, 3, 1, 31, 2)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        ksize = self.get("ksize")
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (ksize, ksize))
        return cv2.morphologyEx(image, cv2.MORPH_CLOSE, kernel)


class MorphGradientFilter(ImageFilter):
    @property
    def name(self) -> str: return "Morphological Gradient"
    @property
    def category(self) -> str: return "Morphological"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("ksize", "Kernel Size", SettingType.INTEGER, 3, 1, 31, 2)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        ksize = self.get("ksize")
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (ksize, ksize))
        return cv2.morphologyEx(image, cv2.MORPH_GRADIENT, kernel)


class TopHatFilter(ImageFilter):
    @property
    def name(self) -> str: return "Top Hat"
    @property
    def category(self) -> str: return "Morphological"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("ksize", "Kernel Size", SettingType.INTEGER, 9, 1, 101, 2)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        ksize = self.get("ksize")
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (ksize, ksize))
        return cv2.morphologyEx(image, cv2.MORPH_TOPHAT, kernel)


class BlackHatFilter(ImageFilter):
    @property
    def name(self) -> str: return "Black Hat"
    @property
    def category(self) -> str: return "Morphological"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("ksize", "Kernel Size", SettingType.INTEGER, 9, 1, 101, 2)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        ksize = self.get("ksize")
        kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (ksize, ksize))
        return cv2.morphologyEx(image, cv2.MORPH_BLACKHAT, kernel)
